File: Java Implementation/MergeSort.java
import java.util.*;

public class MergeSort
{
  public static void main(String... args)
  {   int size;
      Scanner sc = new Scanner(System.in);
      System.out.println("How many datas : "); size = sc.nextInt();
	  ArrayList<Integer> datas= new ArrayList(size);
	  for(int i=0;i<size; i++)
	  {
	     datas.add(sc.nextInt());
	  }
	  System.out.println("Unsorted datas are : ");
	  for(int i=0;i<size; i++)
	  {
	     System.out.print(datas.get(i)+" ,  ");
	  }
	  
	  // Creating mySort Object 
	  MySort<Integer> mySort = new MySort();
	  mySort.sort(datas,size);	  
	  System.out.println("The sorted arrays are : ");	  
	  for(int i=0;i<size;i++) System.out.print(datas.get(i) + ",");
	  
	  MySort<Character> myChars = new MySort();
	  ArrayList<Character> charDatas = new ArrayList(Arrays.asList('j','a','c','b','d','g','e','f','i','h'));
	  myChars.sort(charDatas,size);
	  System.out.println("\nCharacter array : "+ charDatas.toString());
	  
	  MySort<String> myStrs = new MySort();
	  ArrayList<String> myChar = new ArrayList(Arrays.asList("abcde","abcd","b","c","a","d","de","ab","abc","ac"));
	  myStrs.sort(myChar,size);
	  System.out.println("\nString Array: "+ myChar.toString());
		  
	  
	
	 
  }
}

class MySort<T extends Comparable<T>>
{   
   
   void sort(ArrayList<T> datas,int high)
   {    
      // System.out.println("Size of datas : "+ datas.size());
       if((high)<2) 
	   {
		   return;
	   }
	   int mid = high/2;
	   ArrayList<T> left = new ArrayList(mid+1);
	   ArrayList<T> right = new ArrayList(high-mid+1);
	  // System.out.println("\nPrinting inside Merege : ");
	   for(int i=0;i<mid;i++)
	   {
	    left.add(datas.get(i));
	   }
	   for(int i=mid;i<high;i++)
	   {
	    right.add(datas.get(i));
	   }
	   
	   sort(left,mid);
	   sort(right,high-mid);
	   merge(datas,left,mid,right,high-mid);
	   
   }
   
   void merge(ArrayList<T> datas,ArrayList<T> left,int leftSize,ArrayList<T> right,int rightSize)
   {
       int i=0; int j=0;int k=0;
	   
	   while((i<leftSize) && (j<rightSize))
	   {
	       if( left.get(i).compareTo(right.get(j)) < 0)
		   {
		     datas.set(k++,left.get(i++));
		   }
		   else
		   {
		     datas.set(k++,right.get(j++));
		   }
	   }
	   while(i<leftSize)
	   {
	     datas.set(k++,left.get(i++));
	   }
	   	 while(j<rightSize)
	   {
	     datas.set(k++,right.get(j++));
	   }	   
   }
   
}
/*

10
33
10
45
7
98
65
2
65
72
32

*/

File: C++ implementation/MergeSort.cpp
#include<stdio.h>
using namespace std;
void merger(int A[],int left[],int leftSize,int right[],int rightSize)
{   // printf("\nIs it breaks \n");
     int i=0;int j=0; int k=0;
     while(((i<leftSize)&&(j<rightSize)))
     {
         if(left[i]>=right[j]) A[k++] = left[i++];
         else A[k++] = right[j++];
     }
     while(i<leftSize) A[k++] = left[i++];
     while(j<rightSize) A[k++] = right[j++];

     /*printf("\nAfter marge : ");
     for(int i=0; i<(leftSize+rightSize); i++)
     {
         printf("%d, ",A[i]);
     }*/
}

void MySort(int A[],int n)
{    if(n<2) return;
    // printf("\n\nSize is %d : ",n);
     int mid = n/2;
     int left[mid+1]; int right[n-mid];
    // printf("\nLeft : ");
     for(int i=0;i<mid;i++)
     {
         left[i] = A[i]; //printf("%d ,",left[i]);
     }
    // printf("\nRight : ");
     for(int i=mid, j=0; i<n;i++)
     {
         right[j]=A[i]; //printf("%d, ",right[j]);
         j++;
     }
     MySort(left,mid);
     MySort(right,n-mid);
     merger(A,left,mid,right,n-mid);
}
int main()
{
    int n;
    printf("How many Numbers : ");
    scanf("%d",&n); int A[n];
    printf("Total datas are : %d",n);
    printf("Please enter the numbers : ");
    for(int i=0;i<n;i++)
    {
       scanf("%d",&A[i]);
    }
    MySort(A,n);
    printf("\nThe sorted datas are : ");
    for(int i=0;i<n;i++)
    {
       printf("%d, ",A[i]);
    }
    return 0;
}


/*
10
17
8
14
21
4
6
12
7
3
11
*/
